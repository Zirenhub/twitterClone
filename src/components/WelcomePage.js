import {
  MainContainer,
  SideImage,
  MainContent,
  MainTitle,
  MainContentContainer,
  InteractionsContainer,
  InteractionsTitle,
  ButtonsContainer,
  Button,
  LineSeperator,
  ExisitngAccountTitle,
  ExistingAccountContainer,
  TwitterBird,
} from '../styles/WelcomePage.styled';
import googleIcon from '../assets/images/2991148.png';
import appleIcon from '../assets/images/154870.png';

const WelcomePage = () => {
  return (
    <MainContainer>
      <SideImage>
        <TwitterBird></TwitterBird>
      </SideImage>
      <MainContent>
        <MainContentContainer>
          <MainTitle>Happening now</MainTitle>
          <InteractionsContainer>
            <InteractionsTitle>Join Twitter today.</InteractionsTitle>
            <ButtonsContainer>
              <Button bg="#ffffff" color="#0f1419" icon={googleIcon}>
                Sign up with Google
              </Button>
              <Button bg="#ffffff" color="#0f1419" icon={appleIcon}>
                Sign up with Apple
              </Button>
              <LineSeperator>or</LineSeperator>
              <Button bg="#1d9bf0" color="#ffffff">
                Sign up with phone or email
              </Button>
            </ButtonsContainer>
            <ExistingAccountContainer>
              <ExisitngAccountTitle>
                Already have an account?
              </ExisitngAccountTitle>
              <Button bg="transparent" color="#1d9bf0">
                Sign In
              </Button>
            </ExistingAccountContainer>
          </InteractionsContainer>
        </MainContentContainer>
      </MainContent>
    </MainContainer>
  );
};

export default WelcomePage;
