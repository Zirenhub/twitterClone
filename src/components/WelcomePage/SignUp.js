import { useState } from 'react';
import {
  MainContent,
  MainContainer,
  FormContainer,
  CloseButtonContainer,
  CloseButton,
} from '../../styles/WelcomePageStyles/SignUp.styled';

const SignUp = (props) => {
  const { handleClose } = props;

  const [name, setName] = useState('');
  const [password, setPassword] = useState('');
  const [email, setEmail] = useState('');

  const handleNameChange = (e) => {
    setName(e.target.value);
  };

  const handlePasswordChange = (e) => {
    setPassword(e.target.value);
  };

  const handleEmailChange = (e) => {
    setEmail(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
  };

  return (
    <MainContainer>
      <CloseButtonContainer>
        <CloseButton onClick={handleClose}></CloseButton>
      </CloseButtonContainer>

      <MainContent>
        <h1>Create your account</h1>

        <FormContainer onSubmit={handleSubmit}>
          <label>
            Name
            <input type="text" onChange={handleNameChange}></input>
          </label>
          <label>
            Password
            <input type="password" onChange={handlePasswordChange}></input>
          </label>
          <label>
            Email
            <input type="email" onChange={handleEmailChange}></input>
          </label>
          <input type="submit" value="Create Account"></input>
        </FormContainer>
      </MainContent>
    </MainContainer>
  );
};

export default SignUp;
