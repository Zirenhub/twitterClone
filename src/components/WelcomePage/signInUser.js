import { signInWithEmailAndPassword } from 'firebase/auth';
import { auth } from '../../Firebase';

const signInUser = (email, password) => {
  signInWithEmailAndPassword(auth, email, password)
    .then((userCredential) => {
      // Signed in
      const user = userCredential.user;
      // ...
    })
    .catch((error) => {
      const errorCode = error.code;
      const errorMessage = error.message;
      console.log(errorCode);
      console.log(errorMessage);
    });
};

export default signInUser;
