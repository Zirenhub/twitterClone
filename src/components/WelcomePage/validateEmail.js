import validator from 'validator';

const validateEmail = (email) => {
  return validator.isEmail(email);
};

export default validateEmail;
