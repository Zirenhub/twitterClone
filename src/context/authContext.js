import { createContext, useContext, useEffect, useState } from 'react';
import {
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signOut,
  onAuthStateChanged,
  updateProfile,
} from 'firebase/auth';
import { auth } from '../Firebase';

const UserContext = createContext();

const AuthContextProvider = ({ children }) => {
  const [user, setUser] = useState({});

  const createUser = async (email, password, userName) => {
    const createNewUser = await createUserWithEmailAndPassword(
      auth,
      email,
      password
    );
    await updateProfile(auth.currentUser, {
      displayName: userName,
    });

    return createNewUser;
  };

  const signIn = async (email, password) => {
    return signInWithEmailAndPassword(auth, email, password);
  };

  const logout = () => {
    return signOut(auth);
  };

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (currentUser) => {
      setUser(currentUser);
    });

    return () => {
      unsubscribe();
    };
  }, []);

  return (
    <UserContext.Provider value={{ createUser, user, signIn, logout }}>
      {children}
    </UserContext.Provider>
  );
};

const UserAuth = () => {
  return useContext(UserContext);
};

export { AuthContextProvider, UserAuth, UserContext };
