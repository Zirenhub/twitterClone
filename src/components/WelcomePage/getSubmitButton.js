import { SubmitButton } from '../../styles/WelcomePageStyles/SignUp.styled';

const getSubmitButton = (color, value, status) => {
  return (
    <SubmitButton type="submit" color={color} disabled={status}>
      {value}
    </SubmitButton>
  );
};

export default getSubmitButton;
