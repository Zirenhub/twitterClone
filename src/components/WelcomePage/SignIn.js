import {
  CloseButton,
  CloseButtonContainer,
  FormContainer,
  MainContent,
  MainContainer,
  ErrorMessage,
} from '../../styles/WelcomePageStyles/SignUp.styled.js';
import {
  ExtraInformation,
  ResponsiveButtons,
  ResponsiveSignUp,
} from '../../styles/WelcomePageStyles/SignIn.styled.js';
import {
  Button,
  LineSeperator,
} from '../../styles/WelcomePageStyles/WelcomePage.styled.js';
import { useState } from 'react';
import { UserAuth } from '../../context/authContext.js';
import { useNavigate } from 'react-router-dom';
import SignUp from './SignUp.js';
import getSubmitButton from './getSubmitButton.js';
import googleIcon from '../../assets/images/2991148.png';
import appleIcon from '../../assets/images/154870.png';

const SignIn = (props) => {
  const { handleClose, setLoading, currentError, setCurrentError } = props;

  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [showSingUp, setShowSingUp] = useState(false);

  const { signIn } = UserAuth();
  const navigate = useNavigate();

  const handleEmailChange = (e) => {
    setEmail(e.target.value);
  };

  const handlePasswordChange = (e) => {
    setPassword(e.target.value);
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    if (email !== '' && password !== '') {
      try {
        setLoading(true);
        await signIn(email, password);
        navigate('/homepage');
      } catch (error) {
        setCurrentError(error.code);
      }
    }
    setLoading(false);
  };

  const submitButton = (color, status) => {
    return getSubmitButton(color, 'Log In', status);
  };

  const handleSignUp = () => {
    setShowSingUp(true);
  };

  return (
    <>
      {showSingUp ? (
        <SignUp handleClose={handleClose}></SignUp>
      ) : (
        <MainContainer style={{ padding: 20 }}>
          <MainContent>
            <CloseButtonContainer>
              <CloseButton
                data-testid="closeButton"
                onClick={handleClose}
              ></CloseButton>
            </CloseButtonContainer>

            <h1>Sign in to Twitter</h1>

            <ResponsiveButtons>
              <Button bg="#ffffff" color="#0f1419" icon={googleIcon}>
                Sign in with Google
              </Button>
              <Button bg="#ffffff" color="#0f1419" icon={appleIcon}>
                Sign in with Apple
              </Button>
            </ResponsiveButtons>
            <LineSeperator>or</LineSeperator>
            <FormContainer onSubmit={handleSubmit}>
              <label>
                Email
                <input type="email" onChange={handleEmailChange}></input>
              </label>
              <label>
                Password
                <input type="password" onChange={handlePasswordChange}></input>
              </label>
              {email && password
                ? submitButton('#eff3f4', false)
                : submitButton('#505050f0', true)}
            </FormContainer>
            <ExtraInformation>
              <ResponsiveSignUp>
                <p>Don't have an account?</p>
                <button onClick={handleSignUp}>Sign up</button>
              </ResponsiveSignUp>

              {currentError && (
                <ErrorMessage style={{ marginLeft: 'auto' }}>
                  {currentError}
                </ErrorMessage>
              )}
            </ExtraInformation>
          </MainContent>
        </MainContainer>
      )}
    </>
  );
};

export default SignIn;
