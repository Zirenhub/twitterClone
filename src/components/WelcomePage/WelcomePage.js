import {
  MainContainer,
  SideImage,
  MainContent,
  MainTitle,
  MainContentContainer,
  InteractionsContainer,
  InteractionsTitle,
  ButtonsContainer,
  Button,
  LineSeperator,
  ExisitngAccountTitle,
  ExistingAccountContainer,
} from '../../styles/WelcomePageStyles/WelcomePage.styled';
import googleIcon from '../../assets/images/2991148.png';
import appleIcon from '../../assets/images/154870.png';
import twitterLogo from '../../assets/images/iconmonstr-twitter-1.svg';
import { useEffect, useState } from 'react';
import SignUp from './SignUp';
import SignIn from './SignIn';
import { LoadingStyled } from '../../styles/WelcomePageStyles/Loading.styled';
import { UserAuth } from '../../context/authContext';
import { useNavigate } from 'react-router-dom';

const WelcomePage = () => {
  const [showSignUp, setShowSignUp] = useState(false);
  const [showSignIn, setShowSignIn] = useState(false);
  const [loading, setLoading] = useState(false);
  const [currentError, setCurrentError] = useState(null);

  const { user } = UserAuth();
  const navigate = useNavigate();

  useEffect(() => {
    if (user) navigate('/homepage');
  }, []);

  const handleShowSignUp = () => {
    setShowSignUp(true);
  };

  const handleShowSignIn = () => {
    setShowSignIn(true);
  };

  const handleClose = () => {
    setShowSignUp(false);
    setShowSignIn(false);
  };

  useEffect(() => {
    // clear error message when signin or signup page is closed
    return () => {
      setCurrentError(null);
    };
  }, [showSignIn, showSignUp]);

  const Welcome = () => {
    return (
      <MainContentContainer>
        <MainTitle>Happening now</MainTitle>
        <InteractionsContainer>
          <InteractionsTitle>Join Twitter today.</InteractionsTitle>
          <ButtonsContainer>
            <Button bg="#ffffff" color="#0f1419" icon={googleIcon}>
              Sign up with Google
            </Button>
            <Button bg="#ffffff" color="#0f1419" icon={appleIcon}>
              Sign up with Apple
            </Button>
            <LineSeperator>or</LineSeperator>
            <Button bg="#1d9bf0" color="#ffffff" onClick={handleShowSignUp}>
              Sign up with phone or email
            </Button>
          </ButtonsContainer>
          <ExistingAccountContainer>
            <ExisitngAccountTitle>
              Already have an account?
            </ExisitngAccountTitle>
            <Button bg="transparent" color="#1d9bf0" onClick={handleShowSignIn}>
              Sign In
            </Button>
          </ExistingAccountContainer>
        </InteractionsContainer>
      </MainContentContainer>
    );
  };

  return (
    <>
      {loading ? (
        <LoadingStyled>Loading</LoadingStyled>
      ) : (
        <MainContainer>
          <SideImage>
            <img src={twitterLogo} alt="twitter logo"></img>
          </SideImage>
          <MainContent>
            {showSignUp && (
              <SignUp
                handleClose={handleClose}
                setLoading={setLoading}
                currentError={currentError}
                setCurrentError={setCurrentError}
              ></SignUp>
            )}
            {showSignIn && (
              <SignIn
                handleClose={handleClose}
                setLoading={setLoading}
                currentError={currentError}
                setCurrentError={setCurrentError}
              ></SignIn>
            )}
            {!showSignUp && !showSignIn && Welcome()}
          </MainContent>
        </MainContainer>
      )}
    </>
  );
};

export default WelcomePage;
