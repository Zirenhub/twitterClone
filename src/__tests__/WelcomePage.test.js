import '@testing-library/jest-dom';
import { cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import WelcomePage from '../components/WelcomePage/WelcomePage';
import { AuthContextProvider } from '../context/authContext';

afterEach(cleanup);

const user = userEvent.setup();

describe('WelcomePage renders', () => {
  test('Matches snapshot', () => {
    const dom = renderer.create(<WelcomePage />).toJSON();

    expect(dom).toMatchSnapshot();
  });

  test('Main content renders', () => {
    render(<WelcomePage />);

    const signUp = screen.getByRole('button', {
      name: 'Sign up with phone or email',
    });
    const signIn = screen.getByRole('button', {
      name: 'Sign In',
    });

    expect(screen.getByText('Happening now')).toBeInTheDocument();
    expect(signUp).toBeInTheDocument();
    expect(signIn).toBeInTheDocument();
  });
});

describe('Sign up and Sign in windows render', () => {
  test('Sign up form', async () => {
    render(
      <BrowserRouter>
        <AuthContextProvider>
          <WelcomePage />
        </AuthContextProvider>
      </BrowserRouter>
    );

    const signUp = screen.getByRole('button', {
      name: 'Sign up with phone or email',
    });

    await user.click(signUp);

    expect(screen.getByText('Create your account')).toBeInTheDocument();
    expect(screen.queryByText('Happening now')).not.toBeInTheDocument();

    const createAccountButton = screen.getByRole('button', {
      name: 'Create Account',
    });

    expect(createAccountButton).toBeDisabled();

    const nameForm = screen.getByLabelText('Name');

    expect(
      screen.getByText('Name has to be longer than 5 or 5 characters')
    ).toBeInTheDocument();
    await user.type(nameForm, 'ThisIsATest');
    expect(
      screen.queryByText('Name has to be longer than 5 or 5 characters')
    ).not.toBeInTheDocument();
    expect(createAccountButton).toBeDisabled();

    const passwordForm = screen.getByLabelText('Password');

    expect(
      screen.getByText('Password has to be longer than 6 or 6 characters')
    ).toBeInTheDocument();

    await user.type(passwordForm, '12345678910111213');

    expect(
      screen.getByText("Password can't be longer than 16 characters")
    ).toBeInTheDocument();
    expect(
      screen.queryByText('Password has to be longer than 6 or 6 characters')
    ).not.toBeInTheDocument();

    await user.clear(passwordForm);
    await user.type(passwordForm, '12345678');

    expect(
      screen.queryByText('Password has to be longer than 6 or 6 characters')
    ).not.toBeInTheDocument();
    expect(
      screen.queryByText("Password can't be longer than 16 characters")
    ).not.toBeInTheDocument();
    expect(createAccountButton).toBeDisabled();

    const emailForm = screen.getByLabelText('Email');

    expect(screen.getByText('Please enter valid email')).toBeInTheDocument();

    await user.type(emailForm, 'test@');

    expect(screen.getByText('Please enter valid email')).toBeInTheDocument();

    await user.clear(emailForm);
    await user.type(emailForm, 'test@test.com');

    expect(
      screen.queryByText('Please enter valid email')
    ).not.toBeInTheDocument();

    expect(createAccountButton).not.toBeDisabled();

    await user.click(screen.getByTestId('closeButton'));

    expect(screen.queryByText('Create your account')).not.toBeInTheDocument();
    expect(screen.getByText('Happening now')).toBeInTheDocument();
  });

  test('Sign in form', async () => {
    render(
      <BrowserRouter>
        <AuthContextProvider>
          <WelcomePage />
        </AuthContextProvider>
      </BrowserRouter>
    );

    const signIn = screen.getByRole('button', {
      name: 'Sign In',
    });

    await user.click(signIn);

    expect(screen.getByText('Sign in to Twitter')).toBeInTheDocument();
    expect(screen.queryByText('Happening now')).not.toBeInTheDocument();

    const createAccountButton = screen.getByRole('button', {
      name: 'Log In',
    });

    expect(createAccountButton).toBeDisabled();

    // no email or password validation in this component

    const emailForm = screen.getByLabelText('Email');

    await user.type(emailForm, 'test@test.com');
    expect(createAccountButton).toBeDisabled();

    const passwordForm = screen.getByLabelText('Password');
    await user.type(passwordForm, '123123123');
    expect(createAccountButton).not.toBeDisabled();

    await user.click(screen.getByTestId('closeButton'));

    expect(screen.queryByText('Sign in to Twitter')).not.toBeInTheDocument();
    expect(screen.getByText('Happening now')).toBeInTheDocument();
  });

  test('From Sign in to Sign up', async () => {
    render(
      <BrowserRouter>
        <AuthContextProvider>
          <WelcomePage />
        </AuthContextProvider>
      </BrowserRouter>
    );
    const signIn = screen.getByRole('button', {
      name: 'Sign In',
    });

    await user.click(signIn);

    const signUpButton = screen.getByRole('button', {
      name: 'Sign up',
    });

    await user.click(signUpButton);

    expect(screen.queryByText('Sign in to Twitter')).not.toBeInTheDocument();
    expect(screen.queryByText('Happening now')).not.toBeInTheDocument();

    expect(screen.getByText('Create your account')).toBeInTheDocument();
  });
});
