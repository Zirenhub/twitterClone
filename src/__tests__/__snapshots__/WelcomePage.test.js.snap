// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WelcomePage renders Matches snapshot 1`] = `
<div
  className="sc-bcXHqe fuxYUk"
>
  <div
    className="sc-gswNZR gxhXOa"
  >
    <img
      alt="twitter logo"
      src="iconmonstr-twitter-1.svg"
    />
  </div>
  <div
    className="sc-dkrFOg iFTCfC"
  >
    <div
      className="sc-hLBbgP kwRa-DG"
    >
      <h1
        className="sc-eDvSVe gcHYeJ"
      >
        Happening now
      </h1>
      <div
        className="sc-jSUZER fRZfL"
      >
        <h1
          className="sc-gKPRtg cLgLpW"
        >
          Join Twitter today.
        </h1>
        <div
          className="sc-iBYQkv iLdIPl"
        >
          <button
            className="sc-pyfCe cGWgFA"
            color="#0f1419"
          >
            Sign up with Google
          </button>
          <button
            className="sc-pyfCe jZSVON"
            color="#0f1419"
          >
            Sign up with Apple
          </button>
          <div
            className="sc-ftTHYK hIifYD"
          >
            or
          </div>
          <button
            className="sc-pyfCe hrkuLO"
            color="#ffffff"
            onClick={[Function]}
          >
            Sign up with phone or email
          </button>
        </div>
        <div
          className="sc-kDvujY kFtFth"
        >
          <p
            className="sc-jrcTuL bmqUfd"
          >
            Already have an account?
          </p>
          <button
            className="sc-pyfCe bxtpT"
            color="#1d9bf0"
            onClick={[Function]}
          >
            Sign In
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
